x/mongo/driver/topology/server.go

```diff
@@ -290,6 +292,34 @@ func (s *Server) ProcessHandshakeError(err error, startingGenerationNumber uint6
 		return
 	}
 
+	isTimeout := func(err error) bool {
+		for err != nil {
+			if netErr, ok := err.(net.Error); ok {
+				if netErr.Timeout() {
+					return true
+				}
+			}
+			// Handle the case where an error has been replaced by "net.errCanceled", which isn't
+			// exported and can't be compared directly. In this case, just compare the error message.
+			if err.Error() == "operation was canceled" {
+				return true
+			}
+			if wrapper, ok := err.(interface{ Unwrap() error }); ok {
+				err = wrapper.Unwrap()
+			} else {
+				break
+			}
+		}
+
+		return false
+	}
+
+	// Ignore errors that indicate a client-side timeout occurred when using an operation-scoped
+	// deadline (i.e. not using connectTimeoutMS as the connection timeout).
+	if (ctxErr == context.DeadlineExceeded || ctxErr == context.Canceled) && isTimeout(wrappedConnErr) {
```

Comment:
To check my understanding, is the intent behind the `isTimeout(wrappedConnErr)` to prevent a race?

I am not sure if this is realistic, but I imagine the following scenario where this is desirable:
- The context deadline expires just after a socket write error occurs
- The `ctxErr` is `context.DeadlineExceeded` but `wrappedConnErr` is the result of the socket write error.
- Because the socket write error is not a timeout, this proceeds to clear the pool.